File: LinkedList.cpp
#include <iostream>
#include <vector> 

using namespace std;

//Single linked list practice
class lList{
    private:
    lList* next;
    int key;
    int value;

    public:
    lList(int newKey = 0, int newValue = 0){
        key = newKey;
        value = newValue;
        next = NULL;
    }

    ~lList(){
        while(this->getNext()){
            delete this->getNext();
        }
    }

    lList* getNext(){
        return this->next;
    }

    int getKey(){
        return this->key;
    }

    int getValue(){
        return this->value;
    }

    void setNext(lList* newItem){
        this->next = newItem;
    }

    int setKey(int newKey){
        this->key = newKey;
    }

    int setValue(int newValue){
        this->value = newValue;
    }

    void insert(int key, int value){
        lList* cur = this;
        while(cur->getNext()){
            cur = cur->getNext();
        }
        lList* newItem = new lList(key, value);
        cur->setNext(newItem);
    }

    int search(int key){
        lList* cur = this;
        while(cur->getKey() != key && cur->getNext()){
            cur = cur->getNext();
        }
        return ((cur->getKey() == key) ? cur->value : -1);
    }

    void printList(){
        lList* cur = this;
        while(cur){
            cout << "key: " <<  cur->getKey() <<" Value: " << cur->getValue() << endl;
            cur = cur->getNext();
        }
    }

    int deleteKey(int key);
    void deleteDups();
    int kFromLast(int n);
};
lList* head = new lList();

int lList::deleteKey(int key){
    lList* cur = this;
    lList* next = this->getNext();
    
    if(cur->getKey() == key){
        lList* temp = head;
        temp->setNext(NULL);
        delete temp;
        head = next;
        return 0;
    }

    while(next->getKey() != key && next->getNext()){
        cur = cur->getNext();
        next = cur->getNext();
    }
    if(next->getKey() == key){
        cur->setNext(next->getNext());
        next->setNext(NULL);
        delete next;
        return 0;
    }else{
        return -1;
    }
}

void lList::deleteDups(){
    lList* slow = this;
    lList* temp;
    lList* fast = slow ->getNext();
    while(slow){
        while(fast){
            if(slow->getValue() == fast->getValue()){
                temp = fast;
                fast = fast->getNext();
                head->deleteKey(temp->getKey());
            }else{
                fast = fast->getNext();
            }
        }
        if(slow->getNext())
            slow = slow->getNext();
        else break;
        if(slow->getNext())
            fast = slow->getNext();
    }
}

int lList::kFromLast(int n){
    lList* cur = head;
    lList* loc = head;
    int counter = -n;
    while(cur){
        if(counter < 0){
            counter++;
        }else{
            counter = loc->getKey();
            loc = loc->getNext(); 
        }
        cur = cur->getNext();
    }
    return counter;
}

int main(int argc, char* argv []){
    head ->insert(1,50);
    head ->insert(4,50);
    head ->insert(6,50);
    head->insert(2,10);
    head->insert(3,40);
    head->insert(6,10);
    head->insert(7,10);

    
    // cout << head->search(3) << endl;
    cout << head->kFromLast(10) << endl;
    // head->deleteDups();

    // head->printList();
    return 0;
}

File: LinkedList.py
class node:
    def __init__(self, data = None):
        self.__data = data
        self.__next = None
    def getData(self):
        return self.__data
    def setData(self, data):
        self.__data = data
    def getNext(self):
        return self.__next
    def setNext(self, next):
        self.__next = next


class linkedList:
    def __init__(self, data = None):
        self.__head = node(data)
    
    def insert(self, data):
        cur = self.__head
        if(self.__head.getData() is None):
            self.__head.setData(data)
        while(cur.getNext() is not None):
            cur = cur.getNext()
        cur.setNext(node(data))
    
    def search(self, data):
        cur = self.__head
        while(cur is not None):
            if(cur.getData() is data):
                print("Found: ", data)
                return True
            cur = cur.getNext()
        print(data, "Not Found!")
        return False
    
    def print(self):
        cur = self.__head
        print("This is the list:")
        while(cur is not None):
            print("\tNode Value: ", cur.getData())
            cur = cur.getNext()
        print("\n")
        
    def kFromLast(self, k):
        pos = abs(k)
        loc = None
        cur = self.__head
        while cur is not None:
            if pos is not 0:
                pos -= 1
            else:
                if loc is None:
                    loc = self.__head
                else:
                    loc = loc.getNext()
            cur = cur.getNext()
        if loc is None:
            print("K is outside Range!")
            return False
        else:
            print("k from the last is: ", loc.getData())
            return True


    def printBackwards(self):
        cur = self.__head
        array = []
        print("This is the list backwards:")
        while(cur is not None):
            array.append(cur.getData())
            cur = cur.getNext()
        while(len(array) is not 0):
            print("\tNode Value: ", array.pop())
        print("\n")
    
    def remove(self, data):
        cur = self.__head
        if cur is None:
            print("List has no Items!")
            return False
        if cur.getData() is data:
            if cur.getNext() is not None:
                self.__head = cur.getNext()
            print("Removed Item: ", data)
            del cur
            return True
        
        prev = cur
        cur = cur.getNext()
        while cur is not None:
            if cur.getData() is data:
                prev.setNext(cur.getNext())
                del cur
                print("Removed Item: ", data)
                return True
            prev = cur
            cur = cur.getNext()
        print("Can't remove what doesn't exist!")

llist = linkedList(1)
llist.insert(2)
llist.insert(5)
llist.insert(4)
llist.insert(3)
llist.insert(10)
llist.insert(11)
llist.insert(8)

llist.search(10)
llist.remove(10)
llist.search(10)
llist.print()
llist.printBackwards()
llist.kFromLast(-3)